src/transaction/log_recovery_redo_parallel.hpp

```diff
@@ -408,16 +411,63 @@ namespace cublog
       // by design pointer is guaranteed to outlive this instance
       const log_lsa *const m_end_redo_lsa;
       const log_reader::fetch_mode m_log_reader_page_fetch_mode;
+      reusable_jobs_stack *const m_reusable_job_stack;
 
       LOG_RECTYPE m_log_rtype;
   };
 
 
+  /* a preallocated pool of jobs
+   */
+  class reusable_jobs_stack final
+  {
+    public:
+      reusable_jobs_stack ();
+      reusable_jobs_stack (const reusable_jobs_stack &) = delete;
+      reusable_jobs_stack (reusable_jobs_stack &&) = delete;
+
+      ~reusable_jobs_stack ();
+
+      reusable_jobs_stack &operator = (const reusable_jobs_stack &) = delete;
+      reusable_jobs_stack &operator = (reusable_jobs_stack &&) = delete;
+
+      void initialize (std::size_t a_stack_size, const log_lsa *a_end_redo_lsa,
+		       bool force_each_page_fetch);
+      /* used for unit testing */
+      void initialize (std::size_t a_stack_size,
+		       std::function<redo_parallel::redo_job_base * ()> a_job_factory);
+
+      std::size_t size () const
+      {
+	return m_stack_size;
+      }
+
+      redo_parallel::redo_job_base *blocking_pop ();
+      void push (redo_parallel::redo_job_base *a_job);
```

Comment:
Change the type of jobs in reusable_jobs_stack from base to impl and avoid the dynamic_cast.
```suggestion
      redo_parallel::redo_job_impl *blocking_pop ();
      void push (redo_parallel::redo_job_impl *a_job);
```

If this is a problem for the unit test, better to leave reusable_jobs_stack out of the current unit test.

Reply:
Done